custom/icds_reports/utils/aggregation_helpers/distributed/agg_dashboard_acitivity.py

```diff
@@ -0,0 +1,226 @@
+        dashboard_uname_rx = re.compile(r'^\d*\.[a-zA-Z]*@.*')
+
+        return {
+            uname
+            for id, uname in all_users
+            if dashboard_uname_rx.match(uname)
+        }
+
+    def get_user_locations(self):
+        user_locations = list()
+        for username in self.get_dashboard_users:
+            user = CommCareUser.get_by_username(username)
+            loc = user.sql_location
+
+            state_id, district_id, block_id, user_level = None, None, None, None
+
+            if loc and loc.location_type.name == 'state':
+                state_id = loc.location_id
+                district_id = 'All'
+                block_id = 'All'
+                user_level = 1
+            elif loc and loc.location_type.name == 'district':
+                district_id = loc.location_id
+                state_id = loc.get_ancestor_of_type('state').location_id
+                block_id = 'All'
+                user_level = 2
+            elif loc and loc.location_type.name == 'block':
+                block_id = loc.location_id
+                district_id = loc.get_ancestor_of_type('district').location_id
+                state_id = loc.get_ancestor_of_type('state').location_id
+                user_level = 3
+            user_locations.append((
+                username,
+                state_id,
+                district_id,
+                block_id,
+                user_level
+            ))
+
+        return user_locations
+
+    @property
+    def tablename(self):
+        return "{}_{}".format(self.aggregate_parent_table, self.date.strftime("%Y-%m-%d"))
+
+    def drop_table_query(self):
+        return 'DROP TABLE IF EXISTS "{}"'.format(self.tablename)
+
+    def create_table_query(self):
+        return """
+        CREATE TABLE IF NOT EXISTS "{tablename}" (
+            CHECK (date = DATE %(date)s),
+            LIKE "{parent_tablename}" INCLUDING DEFAULTS INCLUDING CONSTRAINTS INCLUDING INDEXES
+        ) INHERITS ("{parent_tablename}")
+        """.format(
+            parent_tablename=self.aggregate_parent_table,
+            tablename=self.tablename,
+        ), {
+            "date": self.date.strftime("%Y-%m-%d"),
+        }
+
+    def get_last_agg_date(self):
+        from custom.icds_reports.models.aggregate import DashboardUserActivityReport
+        result = DashboardUserActivityReport.objects.filter(date__lt=self.date.strftime("%Y-%m-%d")).aggregate(Max('date'))
+        if result:
+            return result['date__max']
+        return None
+
+    def add_missing_users(self):
+        return """
+        INSERT INTO "{tablename}" (
+            username, state_id, district_id,block_id,
+            user_level,date
+        )
+        VALUES
+        """.format(
+            tablename=self.tablename
+        ), {
+            'values': [tuple(list(loc) + [self.date]) for loc in self.get_user_locations()],
+        }
+
+    def rollover_previous_data(self):
```

Comment:
Instead of doing a query to find last agg and then using it down here, you should be able to do it in one query with a window function, or better yet just taking the last value for each column since all users are always updated, much like we do in the other aggs.

Reply:
Hey, I also need last_agg date in the query at the last, that is why I used the same here as well instead of using window function.